pkg/k8s/synced/resources.go

```diff
@@ -117,3 +140,60 @@ func (r *Resources) WaitForCacheSync(resourceNames ...string) {
 		}
 	}
 }
+
+// poll period for underlying client-go wait for cache sync.
+const syncedPollPeriod = 100 * time.Millisecond
+
+// WaitForCacheSyncWithTimeout waits for K8s resources represented by resourceNames to be synced.
+// For every resource type, if an event happens after starting the wait, the timeout will be pushed out
+// to be the time of the last event plus the timeout duration.
+func (r *Resources) WaitForCacheSyncWithTimeout(timeout time.Duration, resourceNames ...string) error {
+	wg := &sync.WaitGroup{}
+	errs := make(chan error)
+	for _, resource := range resourceNames {
+		done := make(chan struct{})
+		wg.Add(1)
+		go func(resource string) {
+			r.WaitForCacheSync(resource)
+			close(done)
+		}(resource)
+
+		go func(resource string) {
```

Comment:
There's nothing wrong with this function, however I am curious if the entire thing gets simpler if you use a `context.WithDeadline()` which is bounced forward on event updates?

Reply:
Good idea, I tried it out this morning. The weird thing about this case is that I have to reset my deadline while waiting on completion.
WithDeadline seems more suited when you have multiple calls imposing deadlines prior to invocation, in which case it just chooses the shortest one.
I'm not sure this makes too much sense just using it to replace inctimer.After since its basically the same code except with having to worry about cleaning up cancels().

I also thought about changing the API a bit to be something like `WaitForCacheSyncWithTimeout(context.Context, ...)`but that doesn't really make sense to me either since there is no guaranteed set deadline when you call the function.

<details><summary>Example</summary>
<p>

```go
ctx, cancel := context.WithTimeout(context.Background(), currTimeout)
defer cancel()
for {
    // Wait until after timeout ends or sync is completed.
    // If timeout is reached, check if an event occurred that would
    // have pushed back the timeout and wait for that amount of time.
    select {
    case <-ctx.Done():
        cancel()
        now := time.Now()
        lastEvent, never := r.getTimeOfLastEvent(resource)
        if never {
            return fmt.Errorf("timed out after %s, never received event for resource %q", timeout, resource)
        }
        if now.After(lastEvent.Add(timeout)) {
            return fmt.Errorf("timed out after %s since receiving last event for resource %q", timeout, resource)
        }
        // We reset the timer to wait the timeout period minus the
        // time since the last event.
        currTimeout = timeout - time.Since(lastEvent)
        ctx, cancel = context.WithTimeout(context.Background(), currTimeout)
        log.Debugf("resource %q received event %s ago, waiting for additional %s before timing out", resource, time.Since(lastEvent), currTimeout)
    case <-done:
        log.Debugf("resource %q cache has synced, stopping timeout watcher", resource)
        cancel()
        return nil
    }
}
```
</p>
</snippet>